lib/new_relic/agent/instrumentation/grpc/server/instrumentation.rb

```diff
@@ -16,18 +16,22 @@ module Server
           INSTANCE_VAR_PORT = :@port_nr
           INSTANCE_VAR_METHOD = :@method_nr
 
-          def handle_with_tracing(active_call, mth, inter_ctx)
+          def handle_with_tracing(active_call, mth, _inter_ctx)
             return yield unless trace_with_newrelic?
 
-            trace_headers = active_call.metadata.delete(NewRelic::NEWRELIC_KEY)
-            ::NewRelic::Agent::DistributedTracing::accept_distributed_trace_headers(trace_headers, 'Other') if ::NewRelic::Agent.config[:'distributed_tracing.enabled']
-
             finishable = NewRelic::Agent::Tracer.start_transaction_or_segment(
               name: mth.original_name,
               category: :web,
               options: server_options(active_call.metadata)
             )
-            yield
+            process_distributed_tracing_headers(active_call)
+
+            begin
+              yield
+            rescue => e
+              NewRelic::Agent.notice_error(e)
+              raise
+            end
```

Comment:
Not something to solve in this issue, but while looking into these lines, I noticed we inconsistently raise the error after calling `NewRelic::Agent.notice_error`. 

These files don't raise:
* lib/new_relic/agent/instrumentation/action_cable_subscriber.rb
* lib/new_relic/agent/instrumentation/sidekiq.rb
* lib/new_relic/agent/instrumentation/sinatra/instrumentation.rb

These files do raise: 
* lib/new_relic/agent/instrumentation/controller_instrumentation.rb
* lib/new_relic/agent/instrumentation/middleware_tracing.rb
* lib/new_relic/agent/instrumentation/active_support_logger/instrumentation.rb
* lib/new_relic/agent/instrumentation/tilt/instrumentation.rb 

A method with a similar purpose, `NewRelic::Agent::Tracer.capture_segment_error`, includes a raise in its definition. Maybe we should follow that pattern unless we truly don't want the raise to occur in some instances? 

I'll transfer this to an issue for further discussion.

Reply:
Issue: https://github.com/newrelic/newrelic-ruby-agent/issues/1290

Reply:
We should definitely follow a consistent pattern. 👍  For the 3 specific examples you gave that don't raise, I have confirmed that it is appropriate for them to not do so. It all comes down to whether or not we are rescuing an exception that would otherwise be left for the instrumented library to handle or not. If we steal an exception like that, we need to re-raise. If we're only operating on a copy of the exception, then no need to re-raise.

Here are specific notes on each of the files you listed:

`lib/new_relic/agent/instrumentation/action_cable_subscriber.rb`
  ^-- this one is finding an exception object in a payload and not actually performing a rescue, so there's no need to re-raise

`lib/new_relic/agent/instrumentation/sidekiq.rb`
  ^-- this one involves a dedicated agent specific error handler that has been registered with Sidekiq. So we're operating on a copy of the underlying exception. The actual exception itself is being handled by a primary Sidekiq error handler and there's no need for our handler to re-raise.

`lib/new_relic/agent/instrumentation/sinatra/instrumentation.rb`
  ^-- this one is operating on a recorded copy of an exception having occured that is stored in an environment hash. We grab the error from the hash and report on it, but we're not actually involved in rescuing the error and don't need to re-raise.

Reply:
Thanks for investigating!